### PR TITLE
feat: Improve RF-DETR with IoU-aware queries, enhanced segmentation, and functional tests

### PR DESCRIPTION
**Pull request by karthikgogisetty**
### 1. IoU-Aware Query Selection & Adaptive Query Allocation
Improves transformer query initialization by:
- Prioritizing higher-IoU spatial locations.
- Dynamically adjusting the number of queries based on feature confidence.
---

### 2. Enhanced Segmentation Head
Upgraded segmentation module adds:
- **Mask quality prediction** (better scoring/ranking).
- **Dynamic mask refinement** using attention over spatial features.

---

### 3. Advanced Data Augmentations (Experimental)
Includes optional training augmentations:
- Mosaic  
- MixUp  
- Copy-Paste  

These augmentations are utilities only and are not enabled by default.

---

## 🔧 Wiring Into Existing Components

Opt-in configuration flags added to:
- `rfdetr/models/transformer.py`
- `rfdetr/models/lwdetr.py`
- `rfdetr/config.py`

Defaults ensure no change to existing model behavior.

---

### Tests
- IoU-aware query selector (shapes, score outputs)
- Adaptive query allocation (range, consistency)
- Enhanced segmentation head (mask logits + quality score shapes)
- ModelConfig with new flags:
  - `use_iou_aware_query`
  - `adaptive_query_allocation`
  - `enhanced_segmentation`
  - `mask_quality_prediction`
  - `dynamic_mask_refinement`

> Note: This functional test suite can be extended with additional cases in future PRs if needed.

### Files
- `functional_testing.py`
- `.github/workflows/functional-tests.yml`

### CI
A new GitHub Actions workflow:
- Installs the package with `uv pip install .`
- Runs `uv run python functional_testing.py` on PRs and key branches.

---

## 🔄 Backward Compatibility

All new features are disabled by default:

| Feature                   | Default |
|---------------------------|---------|
| use_iou_aware_query       | False   |
| adaptive_query_allocation | False   |
| enhanced_segmentation     | False   |
| mask_quality_prediction   | True    |
| dynamic_mask_refinement   | True    |

Existing checkpoints and training/inference scripts remain unchanged.

---

## 📝 Documentation Updates

`README.md` now includes:

### Advanced Features (Experimental)
- IoU-aware query selection  
- Adaptive query allocation  
- Enhanced segmentation head  
- Mask quality prediction  
- Advanced augmentations (Mosaic, MixUp, Copy-Paste)  

### Testing and CI
- Instructions for running:
```
  python functional_testing.py
```

or 

```
uv pip install .
uv run python functional_testing.py
```

### CODE REVIEW

**Comment by karthikgogisetty** [comment]
I have signed the CLA

**Comment by isaacrob** [comment]
Without evidence that these changes improve results I see no reason to merge them

**Comment by anson416** [comment]
Hi. Can we at least have an option to enable [copy-paste](https://arxiv.org/abs/2012.07177) in `TrainConfig`? I am very interested in knowing whether enabling copy-paste would boost training performance for my tiny instance segmentation dataset.

**Comment by Borda** [comment]
> I have signed the CLA

@karthikgogisetty since the check is still pending, could you pls share a printscreen?

**Comment by isaacrob** [comment]
@anson416 not opposed to adding support for additional augs!

**Comment by isaacrob** [comment]
hi! we haven't heard back about evidence that this improves performance so I will suggest that we close this pr. I would recommend that @karthikgogisetty publish a paper if these changes do turn out to make things better :)
